src/plugins/intel_gpu/src/graph/impls/common/loop.cpp

```diff
@@ -37,63 +107,173 @@ struct loop_impl : typed_primitive_impl<loop> {
     void set_node_params(const program_node& arg) override {
         OPENVINO_ASSERT(arg.is_type<loop>());
         const auto& node = arg.as<loop>();
-        _max_iteration = node.get_max_iteration();
         _back_edges = node.get_back_edges();
     }
 
+    void set_body_output_memory(cldnn::network::ptr network, const std::shared_ptr<cldnn::primitive_inst>& inst, memory::ptr mem) const {
+        if (inst->is_input()) {
+            network->set_input_data(inst->id(), mem);
+        } else if (inst->is_output()) {
+            network->set_output_memory(inst->id(), mem);
+        } else {
+            inst->set_output_memory(mem, false);
+        }
+    }
+
+    void change_memory_buffer_static_model(const loop_inst::backedge_memory_mapping& mapping, network::ptr network, int64_t iter) const {
+        if (mapping.type == loop_inst::backedge_memory_mapping::CONCAT_OUTPUT) {
+            if (iter == 0) {
+                set_body_output_memory(network, mapping.to_primitive, mapping.initial_mem);
+            } else if (iter > 0) {
+                auto mem = mapping.concat_mem_mapping->get_sliced_mems().at(iter - 1);
+                set_body_output_memory(network, mapping.to_primitive, mem);
+            } else {
+                OPENVINO_THROW("Invalid iteration count", iter);
+            }
+        } else if (mapping.type ==  loop_inst::backedge_memory_mapping::SINGLE_SHARED) {
+            if (iter == 0) {
+                if (mapping.from_mem != nullptr) {
+                    mapping.from_mem->copy_from(network->get_stream(), *(mapping.initial_mem));
+                }
+            } else {
+                // In dynamic model, output memory is not defined before execution.
+                // After body network execution, replace input memory from initial_mem(external input memory) to output memory.
+                if (mapping.from_mem == nullptr) {
+                    mapping.from_mem = mapping.from_primitive->output_memory_ptr();
+                    OPENVINO_ASSERT(mapping.from_mem != nullptr, "from_mem should not be null");
+                    set_body_output_memory(network, mapping.to_primitive, mapping.from_mem);
+                }
+            }
+        } else if (mapping.type ==  loop_inst::backedge_memory_mapping::SINGLE) {
+            memory::ptr mem1 = mapping.to_primitive->output_memory_ptr();
+            if (iter == 0) {
+                mem1->copy_from(network->get_stream(), *(mapping.initial_mem));
```

Comment:
mem1 is to_mem which will be written by the body network isnt it? should we initialize?

Reply:
Yes, we should initialize mem1 of to_primitive using input node value of outer loop node at first iteration. because to_primitive is the input of body. to_primitive get the data from input node data of loop node and also the output of inner body for previous iteration.